feat: add shortcuts server

rb-determined-ai opened the pull request.

rb-determined-ai committed:
feat: add shortcuts server

stoksc commented on README.md:
```diff
@@ -0,0 +1,21 @@
+
+The server runs in GCP under the name `ci-shortcuts`.  The VM runs with:
+
+ - No service account, so it has no access to GCP APIs
+ - A dedicated VPN, so it can't reach any other Determined instances on
+   internal IP addresses
```
can we terraform this? i will add it if you want.

stoksc replied:
idk, maybe we just wait for an infra person but having this stuff like not floating in the cloud would be nice

rb-determined-ai replied:
who will keep the terraform state files?  That would be my only concern.

rb-determined-ai replied:
Also, my guess is nobody has to update that instance... like ever

rb-determined-ai replied:
update: this will run on the same node as https://det-ci.dzhu.dev

stoksc commented on circleci.go:
```diff
@@ -0,0 +1,344 @@
+	BuildNum  int       `json:"build_num"`
+}
+
+func getURL(url string) ([]byte, error) {
```
why not `http.Get`? http client is meant for reuse (it maintains a connection pool)

stoksc replied:
package level Get uses a singleton http client in the package

rb-determined-ai replied:
I read the [docs](https://pkg.go.dev/net/http#pkg-functions) and they told me to use a client to set the client headers.

I missed the detail about reusing a client object... I can add that.

stoksc commented on circleci.go:
```diff
@@ -0,0 +1,344 @@
+
+type Workflow string
+
+func (w *Workflow) UnmarshalJSON(b []byte) error {
```
the idea here is to use the same model for the db and receiving the json?

rb-determined-ai replied:
Yes.  This seemed less bad than having two separate structs with 80% duplication

stoksc replied:
i think this is a bit error prone, but so is the other way. either way it's fine for this.

stoksc commented on db.go:
```diff
@@ -0,0 +1,191 @@
+
+func OldestUnfinishedBuild(db DBLike) (*Build, error) {
+	var b Build
+	q := db.NewSelect()
+	q = q.Model(&b)
+	q = q.Where("lifecycle != ?", "finished")
+	q = q.Order("build_num ASC")
+	q = q.Limit(1)
```
nit, just use chaining?
```suggestion
	q := db.NewSelect().
		Model(&b).
		Where("lifecycle != ?", "finished").
		Order("build_num ASC").
		Limit(1)
```

rb-determined-ai commented:
@stoksc what did you think of the `DBLike` interface?

I thought it was neat that you could call things on the DB object or on a Tx object, but losing the object-orientedness seems like a high cost, so I'm not sure I'm a fan in the end.

stoksc commented:
yeah, we have something similar for the go master https://github.com/determined-ai/determined/blob/195a5e2e4da4e1359c14f27a45620b5b5b353e22/master/internal/db/postgres_tasks.go#L27 .. it's slightly unfortunate only that it isn't supported in some way in bun already. but it's fine with me.

rb-determined-ai committed:
review feedback

rb-determined-ai commented on db.go:
```diff
@@ -0,0 +1,236 @@
+// Both RootDB and TxDB implement this, but:
+//  - RootDB has a .Begin() method that returns a TxDB
+//  - TxDB has .Commit() and .Rollback() methods
+type DB interface {
+	UpsertBuild(b Build) error
+	GetBuild(buildNum int) (*Build, error)
+	GetBuildsForPull(pull int) ([]Build, error)
+	OldestUnfinishedBuild() (*Build, error)
+	LatestFinishedBefore(before *Build) (*Build, error)
+	GetArchivableBuild() (*Build, error)
+	UnarchivedBuilds() ([]Build, error)
+	ArchiveBuild(buildNum int) error
+	UpsertArtifacts(artifacts []Artifact) error
+	GetArtifacts(buildNum int) ([]Artifact, error)
+}
+
+
+// dbImpl implements our OO-style database interface
+// dbImpl automatically implements DBLike by composition.
+type dbImpl struct {
+	DBLike
+}
+
+// RootDB implements DBLike and dbImpl.
+// Additionally RootDB has a Begin() method.
+type RootDB struct {
+	// bun.DB makes RootDB implement DBLike
+	*bun.DB
+	// dbImpl will just contain another pointer to the *bun.DB
+	dbImpl
+}
+
+// TxDB provides the same OO-style database, plus the usual bun.Tx functions
+// TxDB implements DBLike and dbImple.
+// Additionally it has .Rollback() and .Commit() mechanisms offered by bun.Tx.
+type TxDB struct {
+	*bun.Tx
+	// dbImpl will just contain another pointer to the *bun.Tx
+	dbImpl
+}
```
Ok, I tried harder to get the desired behavior.

I like this a lot.  The only real downside is the need for the interface... but we already have that in the master for other reasons.  If you like this too, I'll likely start with something like this in the master.

rb-determined-ai replied:
going with @ioga's [strategy](https://github.com/determined-ai/ci-shortcuts/pull/2) instead

rb-determined-ai committed:
reconfigure to run behind nginx

rb-determined-ai committed:
switch from Begin to RunInTx

stoksc commented on server.go:
```diff
@@ -0,0 +1,128 @@
+	showURL("model_hub", modelHub, "")
+	showURL("master", master, "#file0")
+	showURL("agent", agent, "#file0")
+	showURL("webui", webui, "")
```
no necessary to change but https://pkg.go.dev/html/template could've been useful

stoksc commented on circleci.go:
```diff
@@ -0,0 +1,319 @@
+		// Our goal is linking to artifacts, which last for 30 days.
+		boundary := time.Now().AddDate(0, 0, -30)
+
+		for offset := 0; true; offset += 100 {
```
it is possible to pass through this loop, and get <100 build but bump the offset by 100. i assume this is fine because if we get <100 the next would have 0 and we would break, but still, seems like it could be a bug.

stoksc commented on circleci.go:
```diff
@@ -0,0 +1,319 @@
+func (c *Client) RefreshBuilds(db RootDB, afterBuildNum int) (error) {
+	return db.RunInTx(func(tx DB) error {
+		for offset := 0; true; offset += 100 {
+			builds, err := c.GetBuilds(100, offset)
```
same here?

stoksc commented on circleci.go:
```diff
@@ -0,0 +1,319 @@
+	})
+}
+
+func (c *Client) RefreshBuilds(db RootDB, afterBuildNum int) (error) {
```
doesn't matter for this but nit refresh/bootstrap dup.

rb-determined-ai replied:
I mean... the duplicated part is mostly boilerplate I think.  I don't see how to refactor it in a way that is more DRY-friendly, without seriously compromising the readability.

stoksc commented on circleci.go:
```diff
@@ -0,0 +1,319 @@
+	}
+}
+
+func (c *Client) RefreshArtifactsThread(db DB, cond *sync.Cond, ready *bool) {
```
really seems like ready and cond should live on the client struct

rb-determined-ai replied:
Eh?  The existence of `Client` is mostly about reusing an `http.Client`.

Maybe I'd be happier putting these Thread functions in server.go.  The cond is for them to talk to each other, and has nothing to do with http.Client calls in general.

stoksc commented on circleci.go:
```diff
@@ -0,0 +1,319 @@
+}
+
+func (c *Client) RefreshBuildsThread(db RootDB, cond *sync.Cond, ready *bool) {
+	setReady := func() {
```
this sync cond seems like a strange way to build a pipeline. just to make sure i 'get' it, could it just be sending on a chan instead?

rb-determined-ai replied:
No, it could not, because it's not a pipeline at all.

The artifact caching caches the most recent cacheable item it can at all times.  If it runs out of stuff to cache, it pauses until the next c.GetBuilds() kicks it (where it checks again and usually pauses again).

Effectively this strategy favors the newest artifacts during a the initial artifact bootstrap, which can take tens of minutes.  Assuming newer PRs are more important than 29-day-old PRs, this means the server is usable a bit sooner.

rb-determined-ai replied:
talked in zoom.  Channels support a "try-send" behavior, so this could be done on a channel with buffer of size 1.

stoksc commented on db.go:
```diff
@@ -0,0 +1,231 @@
+
+// RootDB.RunInTx() wraps bun.DB.RunInTx to provide the DB interface without giving access to the
+// actual bun.Tx object.
+func (db *RootDB) RunInTx(fn func(DB) error) error {
```
i would say when code like this goes into the master, always take the context. we need to start using them.

rb-determined-ai replied:
Agreed, I've already started that in master.

stoksc commented on db.go:
```diff
@@ -0,0 +1,231 @@
+		ExcludeColumn("archived").
+		On("CONFLICT (build_num) DO UPDATE").
+		Set("lifecycle = EXCLUDED.lifecycle")
+	_, err := q.Exec(context.Background())
```
prefer context.TODO or letting the caller decide

stoksc commented on db.go:
```diff
@@ -0,0 +1,231 @@
+
+func (db dbImpl) GetBuild(buildNum int) (*Build, error) {
+	b := Build{BuildNum: buildNum}
+	q := db.NewSelect().
+		Model(&b).
+		WherePK()
+	err := q.Scan(context.Background())
+	if err == sql.ErrNoRows {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, err
+	}
```
chain everything / switch, + elsewhere
```suggestion
	err := db.NewSelect().
		Model(&b).
		WherePK().
		Scan(context.Background())
	switch {
	case err == sql.ErrNoRows:
		return nil, nil
	case err != nil:
		return nil, err
	}
```

rb-determined-ai replied:
what's the point of the switch?

stoksc replied:
style, all error handling in a single block.

rb-determined-ai replied:
didn't take this exact advice, but I did simplify it so to meet the requirement of "one block of error handling"

stoksc commented on db.go:
```diff
@@ -0,0 +1,231 @@
+	}
+	q := db.NewInsert().
+		Model(&artifacts).
+		Ignore()
```
is it possible we may want on conflict set ... instead? seems like artifacts should be immutable but idk circleci's api

rb-determined-ai replied:
I mean... there's only two fields in that database
```sql
		CREATE TABLE IF NOT EXISTS artifacts (
			url TEXT PRIMARY KEY,
			build_num INTEGER NOT NULL
		);
```
I'm comfortable that circleci won't combine url for two artifacts from two different build numbers, so I think this doesn't matter at all.

stoksc commented on db.go:
```diff
@@ -0,0 +1,231 @@
+	var b Build
+	q := db.NewSelect().
+		Model(&b).
+		Where("lifecycle == ?", "finished").
+		Where("archived == false").
```
`Where("lifecycle == 'finished'")` or what is here? what style do we choose (e.g. where below could be written the same way). i think prefer former `Where("archived == ?", false)`, idk.

rb-determined-ai replied:
my rule was roughly "I don't like nested quotes".  But I'm fine settling on the former.

stoksc reviewed (approved):
lgtm. peeked at the server too, neat. where do logs/alerts go for this thing? how do we know if it is in pain.

rb-determined-ai commented:
> where do logs/alerts go for this thing? how do we know if it is in pain

I don't have an answer to that yet

ioga committed:
more buns (#2)

rb-determined-ai committed:
more review feedback

- Bring threads into server.go
- Use more chaining
- Use chan instead of sync.Cond and bool
- Fix bug of using lifecycle == "finished" instead of outcome != nil

rb-determined-ai committed:
add redirects for each type of coverage report

This will reduce clicks from github to desired report from 2 to 1.

rb-determined-ai committed:
fix makefile

rb-determined-ai committed:
update docs
